### DISCUSSION TITLE
Metrics on latency

### DISCUSSION THREAD

**Discussion by ennc0d3** [emqx/emqx#13502] [General]
I used the slow-subscription setting make a note of the slow subscribers, but would be good to see if there is any metrics that i can use.

Are they any metrics that I can use to measure latency? Specifically looking for Qos=1 message the delay between PUB and PUBACK?

Thanks, Naga

**Comment by qzhuyan** [comment]
I don't think so. 
 
Latency is part of perf test not monitoring.

**Comment by zmstone** [comment]
Hi @ennc0d3 
It's quite expensive if we are to report latency for each and every message's passing through.
It's the reason why we have implemented slow-sub to only report the ones above a certain threshold.

We have plan to support latency instrumentation around external calls like authn API call, or query latency.
After the basic functionality is in place, we can consider adding the PUBLISH/PUBACK latency instrumentation.

Preliminary idea is not to report latency histogram, but to use counters instead,
e.g. we allow user to configure the latency buckets, e.g. `["<100ms", "<1s", "<10s", '>=10s']`. 
the corresponding bucket counter is bumped when PUBACK received.
From this, one should be able to compute latency percentiles etc to gain better insight of their traffic pattern.

please let me know if this is something interesting to you.

btw, such features will only be available in Enterprise edition.

**Reply by ennc0d3** [reply]
I was using the slow subscription feature, but when I wanted to confirm its is the sluggish client, I was looking for such metrics on EMQX.
 
 It could be good to have but as of now we are good.
 
 Thanks,
